File: tlg_android/src/ch/bailu/tlg_android/PaintThread.java
package ch.bailu.tlg_android;

import android.content.Context;
import android.util.Log;
import android.view.SurfaceHolder;
import ch.bailu.tlg.InternalContext;

public class PaintThread extends Thread {
    private enum Job {paint, paintAll, none};
    private Job  job = Job.none;
    private AndroidContext tContext;
    private boolean runningThread=false;
    
    private final InternalContext iContext;
    
        
    public PaintThread(InternalContext t, AndroidContext c) {
    	iContext = t;
    	tContext = c;
    }
    
    
    
    public synchronized void run() {
        while(runningThread) waitAndExecuteJob();
    }

    
    private void waitAndExecuteJob() {
        if (haveJob()) {
            executeJob(job);
            jobDone();
        }
        waitForNextJob();
    }

    private void waitForNextJob() {
        try {
            wait();
        } catch (InterruptedException e) {
            runningThread=false;
        }
    }
    
    private boolean haveJob() {
        return (job!=Job.none);
    }
        
    private void executeJob(Job j) {
        if (j==Job.paint) {
            iContext.update(tContext);
        } else {
            iContext.updateAll(tContext);
        }
        tContext.unlockCanvas();
    }


    private void jobDone() {
        job=Job.none;
    }
        
    public void updateAll() {
        requestJob(Job.paintAll);
    }
    
    public void update() {
        requestJob(Job.paint);
    }
    
    private synchronized void requestJob(Job j) {
        if (job != Job.paintAll) job=j;
        notify();
    }

    public synchronized void startPainter(Context context, SurfaceHolder surfaceHolder) {
        if (!runningThread) {
            Log.i("Painter", "Start");
            tContext = new FullGraphicContext(context, surfaceHolder);
            updateAll();
            runningThread=true;
            start();
        }
    }
    
    public void stopPainter() {
        orderShutdown();
    }
    
    private synchronized void orderShutdown() {
        Log.i("Painter", "Stop");
        runningThread=false;
        notify();
    }
}

File: tlg_android/src/main/java/ch/bailu/tlg_android/view/PaintThread.kt
package ch.bailu.tlg_android.view

import android.content.Context
import android.view.SurfaceHolder
import ch.bailu.tlg_android.context.AndroidBaseContext
import ch.bailu.tlg_android.context.AndroidGraphicsContext
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class PaintThread(
    private var tContext: AndroidBaseContext,
    private val cbUpdate: (tContext: AndroidBaseContext)->Unit,
    private val cbUpdateAll: (tContext: AndroidBaseContext)->Unit) : Thread() {

    private enum class Job {
        PAINT, PAINT_ALL, NONE
    }

    private val lock = ReentrantLock()
    private val condition = lock.newCondition()

    private var job = Job.NONE
    private var runningThread = false

    override fun run() {
        while (runningThread) {
            waitAndExecuteJob()
        }
    }

    private fun waitAndExecuteJob() {
        if (haveJob()) {
            executeJob(job)
            jobDone()
        }
        waitForNextJob()
    }

    private fun waitForNextJob() {
        lock.withLock {
            condition.await()
        }
    }

    private fun haveJob(): Boolean {
        return job != Job.NONE
    }

    private fun executeJob(j: Job) {
        if (j == Job.PAINT) {
            cbUpdate(tContext)
        } else {
            cbUpdateAll(tContext)
        }
        tContext.unlockCanvas()
    }

    private fun jobDone() {
        job = Job.NONE
    }

    fun updateAll() {
        requestJob(Job.PAINT_ALL)
    }

    fun update() {
        if (job == Job.NONE) {
            requestJob(Job.PAINT)
        }
    }

    private fun requestJob(j: Job) {
        job = j
        lock.withLock {
            condition.signalAll()
        }
    }

    fun startPainter(context: Context, surfaceHolder: SurfaceHolder) {
        if (!runningThread) {
            tContext = AndroidGraphicsContext(context, surfaceHolder)
            updateAll()
            runningThread = true
            start()
        }
    }

    fun stopPainter() {
        orderShutdown()
    }

    private fun orderShutdown() {
        runningThread = false
        lock.withLock {
            condition.signalAll()
        }
    }
}
